Azure/azure-cli
# [Extension] Support automatically installing an extension if the extension of a command is not installed

**Description**  

When a command failed in `parser._check_value` due to a command word is not in the choice list of its previous word, such as `az vm abc` would fail because `abc` is not a choice under `vm`. This failure could be caused by:
1. Typo, e.g. `az accuont show` or use `get` instead of the correct `show`.
3. The command is from an extension but the extension is not installed.

Previously, we cannot tell the difference and the error message is like this:

> az: 'bluprint' is not in the 'az' command group. See 'az --help'. If the command is from an extension, please make sure the corresponding extension is installed. To learn more about extensions, please visit https://docs.microsoft.com/en-us/cli/azure/azure-cli-extensions-overview

This PR adds a search for the unmatched command in all the extension commands if a close match for the wrong command is not found in all module commands (indicating it's not likely a typo). If the command (not including options) matches a complete extension command, based on config, we can automatically install the extension with or without prompt. Then the command will continue to run by default or users can choose to rerun by themselves.

A mapping of all extension commands to extension names will be fetched from a storage account and stored in ~/.azure/extCmdIndex.json. It is valid for 10 days and will only be refreshed when a parse error happens. An example of the file:
```
{
       "aks": {
                "create": "aks-preview",
                "update": "aks-preview",
                "app": {
                    "up": "deploy-to-azure"
                },
                "use-dev-spaces": "dev-spaces"
        },
        ...
}
```

Configuration:
Both config options are under extension section:
1. use_dynamic_install
yes_prompt: it will use dynamic extension install and prompt to user whether they would like to install the needed extension.
yes_without_prompt: it will install the needed extension directly without confirmation from users.
no: it falls back to previous behavior and just throws an error. This is the default for now.

2. run_after_dynamic_install
yes/true: It’s the default and will continue to run the command after extension is installed.
no/false: It will stop when extension is installed and ask the user to rerun the command.

You can also set with environment variables (AZURE_EXTENSION_USE_DYNAMIC_INSTALL, AZURE_EXTENSION_RUN_AFTER_DYNAMIC_INSTALL)and they have higher priority.

**Testing Guide**  

1.set env var to turn on the feature:
export AZURE_EXTENSION_USE_DYNAMIC_INSTALL=yes_promt
2.run a command from an extension not installed to chekc the behavior.
3.run a real error command to chekc the new clear error message. 


---

This checklist is used to make sure that common guidelines for a pull request are followed.
